File: backend/src/application/use-cases/search/SearchProductsUseCase.ts
import type { SearchProductsUseCasePort } from "@application/ports/incoming/SearchProductsUseCasePort";
import type { ProductCatalogRepository } from "@application/ports/outgoing/ProductCatalogRepository";
import type { BackgroundRefreshQueuePort } from "@application/ports/outgoing/BackgroundRefreshQueuePort";
import type { LoggerPort } from "@application/ports/outgoing/LoggerPort";
import type { IProduct } from "@domain/entities/IProduct";
import type { SearchProductsInput, SearchResult } from "./contracts";

const STALE_THRESHOLD_MS = 24 * 60 * 60 * 1000;
const DEFAULT_PAGE_LIMIT = 20;
const MAX_PAGE_LIMIT = 100;

export class SearchProductsUseCase implements SearchProductsUseCasePort {
	constructor(
		private readonly productCatalogRepository: ProductCatalogRepository,
		private readonly backgroundRefreshQueue: BackgroundRefreshQueuePort,
		private readonly logger: LoggerPort,
	) {}

	async execute(input: SearchProductsInput): Promise<SearchResult> {
		this.logger.info(
			`[SearchProductsUseCase] execute - query: "${input.query}", category: "${input.category}", supermarket: "${input.supermarket}", sortBy: "${input.sortBy}"`,
		);

		const normalizedQuery = input.query?.trim() ?? "";
		const catalogProducts = await this.productCatalogRepository.find({
			query: normalizedQuery,
			category: input.category,
			supermarket: input.supermarket,
			limit: 500,
		});

		const deduplicatedProducts =
			this.deduplicateBySupermarketAndName(catalogProducts);
		const sortedProducts = this.sortByPrice(deduplicatedProducts, input.sortBy);
		const latestScrapedAt = this.findLatestScrapedAt(catalogProducts);
		const refreshDecision = this.computeRefreshDecision(
			normalizedQuery,
			sortedProducts.length,
			latestScrapedAt,
		);

		const isRefreshing = refreshDecision.shouldRefresh
			? this.backgroundRefreshQueue.enqueue(normalizedQuery)
			: false;

		const effectiveLimit = Math.min(input.limit ?? DEFAULT_PAGE_LIMIT, MAX_PAGE_LIMIT);
		const { pageSlice, nextCursor, totalCount } = this.applyPaginationWindow(
			sortedProducts,
			input.cursor,
			effectiveLimit,
		);

		return {
			query: normalizedQuery,
			results: pageSlice,
			warnings: [],
			totalCount,
			nextCursor,
			source: "database",
			scrapedAt: new Date().toISOString(),
			isRefreshing: isRefreshing || undefined,
			refreshReason: isRefreshing ? refreshDecision.refreshReason : undefined,
		};
	}

	private applyPaginationWindow(
		products: IProduct[],
		cursor: string | undefined,
		limit: number,
	): { pageSlice: IProduct[]; nextCursor: string | null; totalCount: number } {
		const totalCount = products.length;
		const parsedOffset = cursor !== undefined ? parseInt(cursor, 10) : 0;
		const offset = Number.isFinite(parsedOffset) && parsedOffset >= 0 ? parsedOffset : 0;
		const pageSlice = products.slice(offset, offset + limit);
		const nextCursor = offset + limit < totalCount ? String(offset + limit) : null;
		return { pageSlice, nextCursor, totalCount };
	}

	private normalizeProductName(name: string): string {
		return name
			.toLowerCase()
			.replace(/[.,/#!$%^&*;:{}=\-_`~()]/g, "")
			.replace(/\s{2,}/g, " ")
			.trim();
	}

	private deduplicateBySupermarketAndName(products: IProduct[]): IProduct[] {
		const uniqueProductsByStoreAndName = new Map<string, IProduct>();

		for (const product of products) {
			const uniqueKey = `${product.supermarket}:${this.normalizeProductName(product.name)}`;
			const existingProduct = uniqueProductsByStoreAndName.get(uniqueKey);
			if (
				!existingProduct ||
				product.pricePerUnit < existingProduct.pricePerUnit
			) {
				uniqueProductsByStoreAndName.set(uniqueKey, product);
			}
		}

		return Array.from(uniqueProductsByStoreAndName.values());
	}

	private sortByPrice(
		products: IProduct[],
		sortBy?: "price_asc" | "price_desc",
	): IProduct[] {
		const sortedProducts = [...products];
		if (sortBy === "price_desc") {
			sortedProducts.sort((left, right) => right.price - left.price);
			return sortedProducts;
		}

		sortedProducts.sort((left, right) => left.price - right.price);
		return sortedProducts;
	}

	private findLatestScrapedAt(products: IProduct[]): Date | null {
		return products.reduce<Date | null>((latestDate, product) => {
			const scrapedAt = new Date(product.scrapedAt);
			if (!latestDate || scrapedAt > latestDate) {
				return scrapedAt;
			}
			return latestDate;
		}, null);
	}

	private computeRefreshDecision(
		normalizedQuery: string,
		productsCount: number,
		latestScrapedAt: Date | null,
	): { shouldRefresh: boolean; refreshReason?: "empty" | "stale" } {
		if (!normalizedQuery) {
			return { shouldRefresh: false };
		}

		if (productsCount === 0) {
			return { shouldRefresh: true, refreshReason: "empty" };
		}

		const hasStaleData =
			latestScrapedAt !== null &&
			Date.now() - latestScrapedAt.getTime() > STALE_THRESHOLD_MS;
		if (hasStaleData) {
			return { shouldRefresh: true, refreshReason: "stale" };
		}

		return { shouldRefresh: false };
	}
}

File: backend-dotnet/src/api/Application/UseCases/Search/SearchProductsUseCase.cs
using Microsoft.Extensions.Logging;
using SharpMonads.Core;
using SupermarketPlanner.Api.Application.DTOs;
using SupermarketPlanner.Api.Application.Ports.Driven;
using SupermarketPlanner.Api.Application.Ports.Driving;
using SupermarketPlanner.Api.Application.Services;
using SupermarketPlanner.Api.Domain;
using SupermarketPlanner.Api.Domain.Services;
using SupermarketPlanner.Api.Domain.ValueObjects;

namespace SupermarketPlanner.Api.Application.UseCases.Search;

public sealed class SearchProductsUseCase(
    IProductCatalogRepository productCatalogRepository,
    IBackgroundRefreshQueue backgroundRefreshQueue,
    ILogger<SearchProductsUseCase> logger) : ISearchProductsUseCase
{
    private const int DefaultPageLimit = 20;
    private const int MaxPageLimit = 100;
    private const string DefaultSource = "database";
    private const string PriceDesc = "price_desc";

    public Task<Either<DomainError, SearchResult>> Invoke(
        SearchProductsInput input,
        CancellationToken cancellationToken)
    {
        logger.LogInformation(
            "[SearchProductsUseCase] Invoke - query: \"{Query}\", category: \"{Category}\", supermarket: \"{Supermarket}\", sortBy: \"{SortBy}\"",
            input.Query, input.Category, input.Supermarket, input.SortBy);

        var normalizedQuery = input.Query?.Trim() ?? string.Empty;
        var effectivePostalCode = input.PostalCode ?? PostalCode.Default.AsString();
        var productCatalogFilters = new ProductCatalogFilters(normalizedQuery, input.Category, input.Supermarket, effectivePostalCode, 500);
        
        return productCatalogRepository
            .Find(productCatalogFilters, cancellationToken)
            .MapAsync(catalogProducts => BuildSearchResult(normalizedQuery, effectivePostalCode, catalogProducts, input));
    }

    private SearchResult BuildSearchResult(
        string normalizedQuery,
        string effectivePostalCode,
        IReadOnlyList<ProductDto> catalogProducts,
        SearchProductsInput input)
    {
        var sorted = SortByPrice(ProductCatalogDeduplicator.Deduplicate(catalogProducts), input.SortBy);
        var refreshDecision = RefreshDecision.Compute(normalizedQuery, sorted.Count, FindLatestScrapedAt(catalogProducts));
        if (refreshDecision.ShouldRefresh)
            backgroundRefreshQueue.Enqueue(normalizedQuery, effectivePostalCode);
        var effectiveLimit = Math.Min(input.Limit ?? DefaultPageLimit, MaxPageLimit);
        var (pageSlice, nextCursor, totalCount) = ApplyPaginationWindow(sorted, input.Cursor, effectiveLimit);

        return new SearchResult(
            Query: normalizedQuery,
            Results: pageSlice,
            Warnings: [],
            TotalCount: totalCount,
            NextCursor: nextCursor,
            Source: DefaultSource,
            ScrapedAt: DateTimeOffset.UtcNow.ToString("o"),
            ActiveRefresh: refreshDecision);
    }

    private static IReadOnlyList<ProductDto> SortByPrice(IReadOnlyList<ProductDto> products, string? sortBy) =>
        sortBy == PriceDesc
            ? [.. products.OrderByDescending(p => p.Price)]
            : [.. products.OrderBy(p => p.Price)];

    private static DateTimeOffset? FindLatestScrapedAt(IReadOnlyList<ProductDto> products) =>
        products.Select(p => (DateTimeOffset?)p.ScrapedAt).DefaultIfEmpty().Max();

    private static (IReadOnlyList<ProductDto> PageSlice, int? NextCursor, int TotalCount) ApplyPaginationWindow(
        IReadOnlyList<ProductDto> products,
        int? cursor,
        int limit)
    {
        var totalCount = products.Count;
        var offset = Math.Max(cursor ?? 0, 0);
        var pageSlice = products.Skip(offset).Take(limit).ToList();
        var nextCursor = offset + limit < totalCount ? offset + limit : (int?)null;
        return (pageSlice, nextCursor, totalCount);
    }
}
